lib/__tests__/callback.test.ts

```diff
@@ -0,0 +1,175 @@
+    beforeEach(() => {
+      navigator.sendBeacon = jest.fn();
+    });
+
+    it("calls the callback with no result", done => {
+      const insightsClient = new AlgoliaAnalytics({
+        requestFn: requestWithSendBeacon
+      });
+      const aa = getFunctionalInterface(insightsClient);
+      aa("init", {
+        appId: process.env.APP_ID,
+        apiKey: process.env.API_KEY
+      });
+      aa("clickedObjectIDsAfterSearch", payload, (error, result) => {
+        expect(error).toBeUndefined();
+        expect(result).toBeUndefined();
+        done();
+      });
+    });
+
+    it("calls the callback with no error from wrong credentials", done => {
+      const insightsClient = new AlgoliaAnalytics({
+        requestFn: requestWithSendBeacon
+      });
+      const aa = getFunctionalInterface(insightsClient);
+      aa("init", {
+        appId: "wrong-app-id",
+        apiKey: "wrong-api-key"
+      });
+      aa("clickedObjectIDsAfterSearch", payload, (error, result) => {
+        expect(result).toBeUndefined();
+        expect(error).toBeUndefined();
+        done();
+      });
+    });
+  });
+
+  describe("XMLHttpRequest", () => {
+    it("calls the callback with result", done => {
+      const insightsClient = new AlgoliaAnalytics({
+        requestFn: requestWithXMLHttpRequest
+      });
+      const aa = getFunctionalInterface(insightsClient);
+      aa("init", {
+        appId: process.env.APP_ID,
+        apiKey: process.env.API_KEY
+      });
+      aa("clickedObjectIDsAfterSearch", payload, (error, result) => {
+        expect(error).toBeNull();
+        expect(result).toEqual('{"status":200,"message":"OK"}');
+        done();
+      });
+    });
+
+    it("calls the callback with error from wrong credentials", done => {
+      const insightsClient = new AlgoliaAnalytics({
+        requestFn: requestWithXMLHttpRequest
+      });
+      const aa = getFunctionalInterface(insightsClient);
+      aa("init", {
+        appId: "wrong-app-id",
+        apiKey: "wrong-api-key"
+      });
+      aa("clickedObjectIDsAfterSearch", payload, (error, result) => {
+        expect(result).toBeUndefined();
+        expect(error).toEqual({
+          responseText: '{"status":401,"message":"Invalid credentials"}',
+          status: 401,
+          statusText: "Unauthorized"
+        });
+        done();
+      });
+    });
+
+    it("calls the callback with error from wrong endpoint", done => {
+      const insightsClient = new AlgoliaAnalytics({
+        requestFn: requestWithXMLHttpRequest
+      });
+      const aa = getFunctionalInterface(insightsClient);
+      aa("init", {
+        appId: "wrong-app-id",
+        apiKey: "wrong-api-key"
+      });
+      insightsClient._endpointOrigin = "https://not_existing_domain!!!.com";
+      aa("clickedObjectIDsAfterSearch", payload, (error, result) => {
+        expect(result).toBeUndefined();
+        expect(error).toEqual({ responseText: "", status: 0, statusText: "" });
```

Comment:
Why don't we have any useful error here?

Reply:
I'm relaying the error from `XMLHttpRequest` and it passes that information in such case.